### PR TITLE
made VS to automatically disable full solution analysis if we detect …

### PR DESCRIPTION
**Pull request by heejaechang**
…low memory for VS.

user can either explicitly turn it back on or re-start VS to re-enable the full solution analysis.

### CODE REVIEW

**Comment by heejaechang** [comment]
related to this PR - https://github.com/dotnet/roslyn/pull/8070

**Comment by heejaechang** [comment]
@dotnet/roslyn-ide @srivatsn @basoundr @mavasani @jmarolf  @KevinH-MS  can you take a look?

**Comment by heejaechang** [comment]
other two issues are known issue tracked by #8079 and #8080

**Comment by heejaechang** [comment]
all three issues are either known or fixed.

**Comment by heejaechang** [comment]
@dotnet-bot retest prtest/mac/dbg/unit32 please

**Comment by heejaechang** [comment]
@dotnet-bot retest prtest/win/dbg/unit32  please

**Comment by heejaechang** [comment]
@dotnet-bot retest prtest/win/dbg/unit64 please

**Comment by heejaechang** [comment]
ping can someone take a look. this is basically same as https://github.com/dotnet/roslyn/pull/8033 except removing the closed file diagnostic option.

File: src/VisualStudio/Core/Def/ServicesVSResources.resx [hunk 1]
```diff
@@ -513,4 +513,7 @@ Use the dropdown to view and switch to other projects this file may belong to.</
   <data name="SynchronizingWithClassView" xml:space="preserve">
     <value>Synchronizing with {0}...</value>
   </data>
+  <data name="FullSolutionAnalysisOff" xml:space="preserve">
+    <value>Full solution analysis disabled due to low memory. go to tools-&gt;options and search "Enable full solution analysis" to turn it back on or restart VS to restore full solution analysis.</value>
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Def/ServicesVSResources.resx]: capital G in "go to".
"and search for"

"and search for "Enable full solution analysis" or restart VS to restore full solution analysis."

--

Can i get a screen shot of what this looks like when it is presented?

File: src/VisualStudio/Core/Def/Implementation/VirtualMemoryNotificationListener.cs [hunk 2]
```diff
@@ -61,6 +68,17 @@ public int OnBroadcastMessage(uint msg, IntPtr wParam, IntPtr lParam)
...
                         _workspaceCacheService.FlushCaches();
+
+                        // turn off full solution analysis
+                        if (_workspace.Options.GetOption(RuntimeOptions.FullSolutionAnalysis))
+                        {
+                            _workspace.Services.GetService<IOptionService>().SetOptions(_workspace.Options.WithChangedOption(RuntimeOptions.FullSolutionAnalysis, false));
+
+                            // let user know full analysis is turned off due to memory concern
+                            // no close info bar action
+                            _workspace.Services.GetService<IErrorReportingService>().ShowErrorInfo(ServicesVSResources.FullSolutionAnalysisOff, () => { });
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Def/Implementation/VirtualMemoryNotificationListener.cs]: why not pass in a callback for 'onEnableClicked' that sets the option back?

THen you also don't need the wonky text about going into "tools options" or needing to restart VS.  The ui would then just be:

"Full solution analysis disabled due to low memory conditions." With an 'enable' button if the user wants to turn it back on right there in the UI.

File: src/VisualStudio/Core/Impl/Options/AbstractOptionPageControl.cs [hunk 3]
```diff
@@ -84,6 +84,18 @@ protected void BindToOption(TextBox textBox, PerLanguageOption<int> optionKey, s
             _bindingExpressions.Add(bindingExpression);
         }
 
+        protected void BindToFullSolutionAnalysisOption(CheckBox checkbox, string languageName)
+        {
+            Binding binding = new Binding();
+
+            binding.Source = new FullSolutionAnalysisOptionBinding(OptionService, languageName);
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Impl/Options/AbstractOptionPageControl.cs]: object intializer?  ```new Binding { Source = ... }```

File: src/VisualStudio/Core/Impl/Options/FullSolutionAnalysisOptionBinding.cs [hunk 4]
```diff
@@ -0,0 +1,51 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using Microsoft.CodeAnalysis.Options;
+using Microsoft.CodeAnalysis.Shared.Options;
+
+namespace Microsoft.VisualStudio.LanguageServices.Implementation.Options
+{
+    internal class FullSolutionAnalysisOptionBinding
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Impl/Options/FullSolutionAnalysisOptionBinding.cs]: Is this a ThreadSffinitizedType?

File: src/VisualStudio/Core/Def/Implementation/VirtualMemoryNotificationListener.cs (hunk 2 above)
Comment (heejaechang) [inline src/VisualStudio/Core/Def/Implementation/VirtualMemoryNotificationListener.cs]: I thought about that before and choose not to do it since they will certainly OOM. this option change will let diagnostic service to throw away any data we no longer needed.

the reason I gave users an option to turn this back on is if user wants to working on other solution, they can go ahead without restarting VS by set those options back on.

File: src/VisualStudio/Core/Impl/Options/AbstractOptionPageControl.cs (hunk 3 above)
Comment (heejaechang) [inline src/VisualStudio/Core/Impl/Options/AbstractOptionPageControl.cs]: sure.

File: src/VisualStudio/Core/Impl/Options/FullSolutionAnalysisOptionBinding.cs (hunk 4 above)
Comment (heejaechang) [inline src/VisualStudio/Core/Impl/Options/FullSolutionAnalysisOptionBinding.cs]: not sure, but this is for WPF, so probably only used in FG.

File: src/VisualStudio/Core/Def/Implementation/VirtualMemoryNotificationListener.cs (hunk 2 above)
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Def/Implementation/VirtualMemoryNotificationListener.cs]: > I thought about that before and choose not to do it since they will certainly OOM

But in your approach you then tell them how to re-enable things, which will then cause an OOM...

>  they can go ahead without restarting VS by set those options back on.

Or they could just click on the button :)

If they OOM, they OOM.  We told them they're in a low memory condition.  If they say "damnit... i still want to go ahead" why would we stop them?  Or, if we think we should stop them, then we shouldn't have a way for them to re-enable it in any way.

--

Also, the appraoch of telling someone "open tools|options, and then search for blah blah blah" is *exceptionally* user-unfriendly.  It's exactly waht people do not like.  "Why are you telling me how to do this when you could just go do it yourself?  This is so annoying for me to have to deal with!"

File: src/VisualStudio/Core/Impl/Options/FullSolutionAnalysisOptionBinding.cs (hunk 4 above)
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Impl/Options/FullSolutionAnalysisOptionBinding.cs]: Please derive from ForegroundThreadAffinitizedObject then and add appropriate asserts.

File: src/VisualStudio/Core/Def/Implementation/VirtualMemoryNotificationListener.cs (hunk 2 above)
Comment (heejaechang) [inline src/VisualStudio/Core/Def/Implementation/VirtualMemoryNotificationListener.cs]: I am not sure whether my English was bad or not. but " on other solution" was the key point. not letting people keep working on same solution by turning the option back on.

File: src/VisualStudio/Core/Def/ServicesVSResources.resx (hunk 1 above)
Comment (heejaechang) [inline src/VisualStudio/Core/Def/ServicesVSResources.resx]: image

File: src/VisualStudio/Core/Def/Implementation/VirtualMemoryNotificationListener.cs (hunk 2 above)
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Def/Implementation/VirtualMemoryNotificationListener.cs]: No no.  I understood the point you're making here.  What i'm saying is that the UI does not convey that properly at all.

You're telling users in the UI "if you want this back, here's how you go turn it on".  There's nothing to tell them "but if you do this, you'll likely OOM".

So since you've already decided you're going to tell them how to do it, you might as well give them the way to do it.  Right now the approach you're taking makes for a *very* user-unfriendly UI.  Telling a user that they need to go into tools|options and then *search* for something, is a *super* unpleasant set of instructions.  It's antithetical to how computers should behave.  

--

We have to decide on a few things:

Do we even want to allow users to re-enable this?  If no, then we don't need a button, and we don't need all that text explaining how to turn it back on.  If yes, then we should go for the much better UI approach.  

If we're worried about them causing an OOM by doing this, then we shouldn't bother giving them the button *or* telling them how to re-enable it.  Right now we've got a really unpleasant approach.  We don't want them to re-enable, unless they're on another solution, but we want htem to have to go through a really unpleasant set of steps to re-enable in that case.  That's not a good user experience *at all*.

File: src/VisualStudio/Core/Def/ServicesVSResources.resx (hunk 1 above)
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Def/ServicesVSResources.resx]: Yeah.  That's super-ick.  How about:

"Low memory detected.  Full solution analysis disabled."

--

also, if we want FullSolutionAnalysis to come back when this solution is closed and another is opened... why not just do that automatically?  We know when that has happened, so we could just automatically do that on behalf of the user.  Then we don't need any buttons.  We're just telling them what has happened.  And once the situation is cleared up, VS is working properly again.

Comment (heejaechang) [inline src/VisualStudio/Core/Def/ServicesVSResources.resx]: ah, that is good idea.

Comment (heejaechang) [inline src/VisualStudio/Core/Def/ServicesVSResources.resx]: done

File: src/VisualStudio/Core/Def/Implementation/VirtualMemoryNotificationListener.cs (hunk 2 above)
Comment (heejaechang) [inline src/VisualStudio/Core/Def/Implementation/VirtualMemoryNotificationListener.cs]: changed wording to "Low memory detected. Full solution analysis disabled for this solution." and it will automatically re-enable full solution analysis if new solution is opened.

File: src/VisualStudio/Core/Impl/Options/FullSolutionAnalysisOptionBinding.cs (hunk 4 above)
Comment (heejaechang) [inline src/VisualStudio/Core/Impl/Options/FullSolutionAnalysisOptionBinding.cs]: it will be only called from UI thread, but object itself doesnt rely on being called from FG. so I dont think we need that assert.

**Comment by heejaechang** [comment]
ping? can I get some sign off?

**Comment by heejaechang** [comment]
same issue as this. https://github.com/dotnet/roslyn/issues/7998

**Comment by heejaechang** [comment]
retest prtest/win/dbg/unit64 please
// Previous failure: http://dotnet-ci.cloudapp.net/job/roslyn_prtest_win_dbg_unit64/3061/
// Retest reason: known issue.

File: src/VisualStudio/Core/Impl/Options/FullSolutionAnalysisOptionBinding.cs [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+            }
+
+            set
+            {
+                var oldOptions = _optionService.GetOptions();
+
+                // set normal option first
+                var newOptions = oldOptions.WithChangedOption(_closedFileDiagnostics, _languageName, value);
+
+                // we only enable this option if it is disabled. we never disable this option here.
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Impl/Options/FullSolutionAnalysisOptionBinding.cs]: Why not?  That seems very weird and asymettric.

**Comment by CyrusNajmabadi** [comment]
I don't see the code the re-enables analysis if a new solution is loaded.  Can you point it out to me?

File: src/VisualStudio/Core/Impl/Options/FullSolutionAnalysisOptionBinding.cs (hunk 4 above)
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Impl/Options/FullSolutionAnalysisOptionBinding.cs]: It's good to put in threading expectations.  That way if something changes, someone can investigate and decide if it is a problem or not.

File: src/VisualStudio/Core/Impl/Options/FullSolutionAnalysisOptionBinding.cs (hunk 5 above)
Comment (heejaechang) [inline src/VisualStudio/Core/Impl/Options/FullSolutionAnalysisOptionBinding.cs]: the runtime option will only be disabled by us. if user explicit disable the option, it will disable user explicit option (CloseFileDiagnostics) and for user end result is same.

we do this so that we have 2 options one controlled by user and persisted, one controlled by us and not persisted.

if user has full analysis on, and we disabled it, user can go and explicit enable it again or if user restart VS, full analysis will be back on.

basically we show one aggregated option to users, and internally it is combination of option user controlling and option we controlling.

File: src/Features/Core/Portable/SolutionCrawler/WorkCoordinator.cs [hunk 6]
```diff
@@ -329,6 +330,9 @@ private void ProcessSolutionEvent(WorkspaceChangeEventArgs e, IAsyncToken asyncT
 
             private void OnSolutionAdded(Solution solution)
             {
+                // first make sure full solution analysis is on.
+                _optionService.SetOptions(solution.Workspace.Options.WithChangedOption(RuntimeOptions.FullSolutionAnalysis, true));
```
Comment (heejaechang) [inline src/Features/Core/Portable/SolutionCrawler/WorkCoordinator.cs]: here.

File: src/VisualStudio/Core/Impl/Options/FullSolutionAnalysisOptionBinding.cs (hunk 4 above)
Comment (heejaechang) [inline src/VisualStudio/Core/Impl/Options/FullSolutionAnalysisOptionBinding.cs]: I believe that is job for caller if they have dependency on FG. this code, it is completely fine, they use it in different code path in BG. it will work just fine.

**Comment by CyrusNajmabadi** [comment]
Ah ok, thanks.

**Comment by CyrusNajmabadi** [comment]
I'm good with this.  But i'd like another IDE senior signoff.  @dotnet/roslyn-ide  can someone take a look?

File: src/VisualStudio/Core/Impl/Options/FullSolutionAnalysisOptionBinding.cs (hunk 5 above)
Comment (srivatsn) [inline src/VisualStudio/Core/Impl/Options/FullSolutionAnalysisOptionBinding.cs]: It's worth adding this to the comment.

**Comment by srivatsn** [comment]
:+1:
